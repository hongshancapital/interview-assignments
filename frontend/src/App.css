.App {
  text-align: center;
  position: relative;
}
