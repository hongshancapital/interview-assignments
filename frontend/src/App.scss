.App {
  text-align: center;
  position: relative;
}

.title {
  font-size: 56px;
  font-weight: 600;
}

.text {
  font-size: 28px;
  font-weight: 400;
}
