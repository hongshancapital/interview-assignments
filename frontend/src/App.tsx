import React from 'react';
import Carousel from './carousel';
import { CarouselData } from './const';

const beforeChange = (from: number, to: number) => {
  console.log('beforeChange', `${from}-${to}`);
};

const afterChange = (current: number) => {
  console.log('afterChange', current);
};

function App() {
  return (
    <Carousel
      duration={3000}
      width={window.document.body.clientWidth}
      height={window.innerHeight}
      beforeChange={beforeChange}
      afterChange={afterChange}
      data={CarouselData}
    />
  );
}

export default App;
