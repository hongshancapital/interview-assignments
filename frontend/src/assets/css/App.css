.App {
  text-align: center;
  position: relative;
}


/*slides*/
.color {
  mix-blend-mode: difference;
  color: #fff;
}

.slides {
  width: 500px;
  height: 300px;
  overflow: hidden;
  margin: auto;
  position: relative;
}

.slides-item {
  display: flex;
  width: 100%;
  height: 100%;
  transition: transform 1s;
}

.slide {
  position: relative;
  width: 100%;
  background-color: #000;
}

.slide-img {
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  object-fit: cover;
}

.slide-title {
  font-size: 24px;
  font-weight: 600;
  white-space: pre-wrap;
  margin: 20px;
}

.slide-text {
  padding: 0 20px;
  margin: 0;
  font-size: 18px;
}

.slides-nav {
  position: absolute;
  bottom: 10px;
  left: 0;
  width: 100%;
  display: flex;
  justify-content: center;
}

.nav-bar {
  width: 30px;
  height: 3px;
  background-color: rgba(0, 0, 0, 0.5);
  margin-right: 5px;
  border-radius: 20px;
  overflow: hidden;
}

.nav-progress {
  background-color: #fff;
  background-blend-mode: difference;
  width: 0;
  height: 100%;
}