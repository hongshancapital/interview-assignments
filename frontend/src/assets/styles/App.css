.App {
  text-align: center;
  position: relative;
}

.title {
  font-size: 56px;
  font-weight: 600;
}

.text {
  font-size: 28px;
  font-weight: 400;
}

.my_carousel_item {
  width: 480px;
  height: 300px;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  position: relative;
}

.my_carousel_item2 {
  width: 960px;
  height: 600px;
}

.my_carousel_item3 {
  width: 1440px;
  height: 900px;
}

.my_carousel_item .description {
  position: absolute;
  text-align: center;
  width: 100%;
  top: 36px;
}

.my_carousel_item._white {
  color: #fff;
}

.my_carousel_item h3 {
  font-size: 26px;
}
.my_carousel_item p {
  font-size: 12px;
}

.my_carousel_item2 h3 {
  font-size: 56px;
}
.my_carousel_item2 p {
  font-size: 28px;
}

.my_carousel_item3 h3 {
  font-size: 86px;
}
.my_carousel_item3 p {
  font-size: 43px;
}
