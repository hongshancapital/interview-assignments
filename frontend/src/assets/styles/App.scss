.App {
  text-align: center;
  position: relative;
}

.title {
  font-size: 56px;
  font-weight: 600;
}

.text {
  font-size: 28px;
  font-weight: 400;
}

.my_carousel_item {
  width: 480px;
  height: 300px;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  position: relative;
  &.my_carousel_item2 {
    width: 960px;
    height: 600px;
    h3 {
      font-size: 56px;
    }
    p {
      font-size: 28px;
    }
  }
  &.my_carousel_item3 {
    width: 1440px;
    height: 900px;
    h3 {
      font-size: 86px;
    }
    p {
      font-size: 43px;
    }
  }

  &._white {
    color: #fff;
  }
  .description {
    position: absolute;
    text-align: center;
    width: 100%;
    top: 36px;
  }

  h3 {
    font-size: 26px;
  }
  p {
    font-size: 12px;
  }
}
