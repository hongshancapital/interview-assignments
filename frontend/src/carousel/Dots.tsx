import { useCallback } from "react";
import styled from "styled-components";

interface IDotsProps {
  count: number;
  duration: number;
  selectedIndex: number;
}

const DOT_WIDTH = 60;

const StyledDots = styled.div<{ duration: number }>`
  display: flex;
  justify-content: center;
  position: absolute;
  width: 100%;
  left: 0;
  bottom: 30px;
  text-align: center;

  .dot {
    background-color: #acacac;
    height: 3px;
    width: ${DOT_WIDTH}px;
    margin: 0 6px;
    border-radius: 2px;
    cursor: pointer;

    .active {
      background-color: white;
      height: 3px;
      animation: ${(props) => `progress ${props.duration}ms linear`};
    }
  }

  @keyframes progress {
    0% {
      width: 0px;
    }
    100% {
      width: ${DOT_WIDTH}px;
    }
  }
`;

export const Dots: React.FC<IDotsProps> = (props) => {
  const { count, duration, selectedIndex } = props;
  const renderDot = useCallback(
    (index: number) => {
      const isActive =
        index === selectedIndex || (selectedIndex === count && index === 0);
      return (
        <div key={index} className="dot">
          <div className={isActive ? "active" : ""} />
        </div>
      );
    },
    [selectedIndex]
  );

  return (
    <StyledDots duration={duration}>
      {new Array(count).fill(0).map((_value, index) => {
        return renderDot(index);
      })}
    </StyledDots>
  );
};
