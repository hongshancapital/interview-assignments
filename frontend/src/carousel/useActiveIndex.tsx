import { useState, useEffect } from 'react';

interface useActiveIndexProps {
  duration?: number;
  count: number;
}

function useActiveIndex(props: useActiveIndexProps) {
  const { duration = 3000, count } = props;
  const [curIndex, setCurIndex] = useState(0);

  useEffect(() => {
    const timer = setInterval(() => {
      const index = (curIndex + 1) % count;
      setCurIndex(index);
    }, duration);
    return () => clearInterval(timer);
  }, [curIndex, duration, count]);

  return { curIndex, setCurIndex };
}

export default useActiveIndex;
