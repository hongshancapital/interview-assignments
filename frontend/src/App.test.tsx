import React from 'react';
import { act, render, screen, fireEvent } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import useActiveIndex from './carousel/useActiveIndex';
import Dots from './carousel/dots';
import App from './App';

jest.useFakeTimers();

describe('carousel component unit testing', () => {
  test('snapshot', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
  test('useActiveIndex hooks', () => {
    const { result } = renderHook(() => useActiveIndex({ count: 3 }));
    expect(result.current.curIndex).toBe(0);

    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(result.current.curIndex).toBe(1);

    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(result.current.curIndex).toBe(2);

    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(result.current.curIndex).toBe(0);

    jest.clearAllTimers();
  });

  test('dot component', async () => {
    const { result } = renderHook(() => useActiveIndex({ count: 3 }));

    const props = {
      count: 3,
      duration: 3000,
      curIndex: 0,
      setCurIndex: result.current.setCurIndex,
    };

    act(() => {
      render(<Dots {...props} />);
    });

    const dot = document.querySelector('.dotContainer');
    expect(dot).toBeInTheDocument();

    const dotBoxs = document.querySelectorAll('.dotBox');
    expect(dotBoxs.length).toBe(3);

    const dotCenter = document.querySelectorAll('.dotCenter.active');
    expect(dotCenter.length).toBe(1);
  });

  test('dot click', async () => {
    render(<App />);

    const carousel = await screen.findByTestId('carousel');
    expect(carousel.classList.contains('active-0')).toBe(true);
    const dot1 = await screen.findByTestId('dot_1');
    fireEvent.click(dot1);
    expect(carousel.classList.contains('active-1')).toBe(true);
  });
});
