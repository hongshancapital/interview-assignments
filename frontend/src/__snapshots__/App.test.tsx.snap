// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`carousel component unit testing snapshot 1`] = `
<div>
  <div
    class="container"
    style="width: 0px; height: 768px;"
  >
    <div
      class="carousel active-0"
      data-testid="carousel"
    >
      <div>
        <div
          class="carousel-item"
          style="background-color: rgb(17, 17, 17); color: rgb(255, 255, 255);"
        >
          <div
            class="carousel-item-text"
          >
            <h1>
              xPhone
            </h1>
            <p>
              Lots to love. Less to spend.
            </p>
            <p>
              Starting at $399.
            </p>
          </div>
          <img
            alt=""
            src="iphone.png"
          />
        </div>
      </div>
      <div>
        <div
          class="carousel-item"
          style="background-color: rgb(250, 250, 250); color: rgb(0, 0, 0);"
        >
          <div
            class="carousel-item-text"
          >
            <h1>
              Tablet
            </h1>
            <p>
              Just the right amount of everything.
            </p>
          </div>
          <img
            alt=""
            src="tablet.png"
          />
        </div>
      </div>
      <div>
        <div
          class="carousel-item"
          style="background-color: rgb(241, 241, 243); color: rgb(0, 0, 0);"
        >
          <div
            class="carousel-item-text"
          >
            <h1>
              Buy a Tablet or xPhone for college.
            </h1>
            <h1>
              Get arPods.
            </h1>
          </div>
          <img
            alt=""
            src="airpods.png"
          />
        </div>
      </div>
    </div>
    <div
      class="dotContainer"
    >
      <div
        class="dotBox"
        data-testid="dot_0"
      >
        <div
          class="dotCenter active"
          style="transition: width 3000ms;"
        />
      </div>
      <div
        class="dotBox"
        data-testid="dot_1"
      >
        <div
          class="dotCenter "
        />
      </div>
      <div
        class="dotBox"
        data-testid="dot_2"
      >
        <div
          class="dotCenter "
        />
      </div>
    </div>
  </div>
</div>
`;
